library/core/src/num/uint_macros.rs

```diff
@@ -3309,6 +3318,18 @@ macro_rules! uint_impl {
             let mut base = self;
             let mut acc = 1;
 
+            if intrinsics::is_val_statically_known(base) {
+                // change of base:
+                // if base == 2 ** k, then
+                //    (2 ** k) ** n
+                // == 2 ** (k * n)
+                // == 1 << (k * n)
+                if base.is_power_of_two() {
+                    let k = base.ilog2();
+                    return 1 << (k * exp)
```

Comment:
I've been staring at this line, and I think it's wrong in release mode.

Take `2_u16.pow(16)`, for example.  Today that gives zero, which is what I'd expect, since that's the infinite-precision result wrapped to `u16`, but the shift would be `1_u16 << (1 * 16)` ⇒ `1_u16 << 16` ⇒ `1_u16`.

So please add some more tests to either show I'm wrong or fix it if I'm right.  Be sure to cover things like `4_u16.pow(0x8000_0000)` too, since the multiplication of the exponent could also overflow.

Reply:
I think you're right. We should use `1 << (k * exp)` when overflow checks are enabled, and `1.unbounded_shl(k * exp)` when overflow checks are disabled. But I don't think there is a way to choose between the two depending on the caller crate's compiler flags. There is [`cfg(overflow_checks)`](github.com/rust-lang/rust/issues/111466) but it sounds like it would always resolve to `false` in the `core` crate